File: models/Gpt-3.5-turbo/cpp/code/problems/geometry/Point_in_Polygon.cpp

#include <iostream>
#include <vector>

using namespace std;

struct Point {
    int x, y;
};

bool isOnBoundary(const Point& p, const Point& a, const Point& b) {
    if (p.x >= min(a.x, b.x) && p.x <= max(a.x, b.x) &&
        p.y >= min(a.y, b.y) && p.y <= max(a.y, b.y)) {
        return true;
    }
    return false;
}

int countIntersections(const Point& p, const Point& a, const Point& b) {
    if (isOnBoundary(p, a, b)) {
        return 0;
    }

    int intersections = 0;

    // Check if the line segment intersects with the ray
    if ((a.y > p.y && b.y <= p.y) || (b.y > p.y && a.y <= p.y)) {
        // Calculate the x-coordinate of the intersection point
        double x = a.x + (double(p.y - a.y) / double(b.y - a.y)) * (b.x - a.x);
        if (x > p.x) {
            intersections++;
        }
    }

    return intersections;
}

string pointLocation(const Point& p, const vector<Point>& polygon) {
    int intersections = 0;
    int n = polygon.size();

    // Check each edge of the polygon
    for (int i = 0; i < n; i++) {
        int j = (i + 1) % n;
        intersections += countIntersections(p, polygon[i], polygon[j]);
    }

    if (intersections % 2 == 1) {
        return "INSIDE";
    } else if (intersections == 0) {
        return "BOUNDARY";
    } else {
        return "OUTSIDE";
    }
}

int main() {
    int n, m;
    cin >> n >> m;

    vector<Point> polygon(n);
    for (int i = 0; i < n; i++) {
        cin >> polygon[i].x >> polygon[i].y;
    }

    for (int i = 0; i < m; i++) {
        Point p;
        cin >> p.x >> p.y;
        cout << pointLocation(p, polygon) << endl;
    }

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/geometry/Point_in_Polygon.py

def orientation(x1, y1, x2, y2, x3, y3):
    # Calculate the orientation of three points (x1, y1), (x2, y2), (x3, y3)
    # Returns 0 if the points are collinear, 1 if clockwise, and 2 if counterclockwise
    val = (y2 - y1) * (x3 - x2) - (y3 - y2) * (x2 - x1)
    if val == 0:
        return 0  # Collinear
    elif val > 0:
        return 1  # Clockwise
    else:
        return 2  # Counterclockwise

def on_boundary(x1, y1, x2, y2, x, y):
    # Check if the point (x, y) lies on the boundary of the line segment (x1, y1) to (x2, y2)
    if x >= min(x1, x2) and x <= max(x1, x2) and y >= min(y1, y2) and y <= max(y1, y2):
        if (y2 - y1) * (x - x1) == (y - y1) * (x2 - x1):
            return True
    return False

def inside_polygon(polygon, x, y):
    # Check if the point (x, y) is inside, outside or on the boundary of the polygon
    n = len(polygon)
    if n < 3:
        return "OUTSIDE"
    
    count = 0
    for i in range(n):
        x1, y1 = polygon[i]
        x2, y2 = polygon[(i + 1) % n]
        
        if on_boundary(x1, y1, x2, y2, x, y):
            return "BOUNDARY"
        
        if y1 <= y and y2 > y or y2 <= y and y1 > y:
            if orientation(x1, y1, x2, y2, x, y) == 0:
                return "BOUNDARY"
            if orientation(x1, y1, x2, y2, x, y) == 1:
                count += 1
    
    if count % 2 == 1:
        return "INSIDE"
    else:
        return "OUTSIDE"

# Read input
n, m = map(int, input().split())
polygon = []
for _ in range(n):
    x, y = map(int, input().split())
    polygon.append((x, y))
points = []
for _ in range(m):
    x, y = map(int, input().split())
    points.append((x, y))

# Process each point
for point in points:
    x, y = point
    result = inside_polygon(polygon, x, y)
    print(result)
